vra/resource_block_device.go

```diff
@@ -249,7 +252,108 @@ func resourceBlockDeviceRead(d *schema.ResourceData, m interface{}) error {
 
 func resourceBlockDeviceUpdate(d *schema.ResourceData, m interface{}) error {
 
-	return fmt.Errorf("Updating a block device resource is not allowed")
+	log.Printf("Starting to update the vra_block_device resource with name %s", d.Get("name"))
+	apiClient := m.(*Client).apiClient
+
+	id := d.Id()
+	if d.HasChange("capacity_in_gb") {
+		resizeDisk(d, apiClient, id)
+	}
+
+	log.Printf("Finished updating vra_block_device resource with name %s", d.Get("name"))
+	return nil
+}
+
+func resizeDisk(d *schema.ResourceData, apiClient *client.MulticloudIaaS, id string) error {
+
+	log.Printf("Starting resize of vra_block_device resource with name %s", d.Get("name"))
+	// get the block device by id
+	resp, err := apiClient.Disk.GetBlockDevice(disk.NewGetBlockDeviceParams().WithID(id))
+	if err != nil {
+		return err
+	}
+	// get the deployment id associated with the block device
+	deploymentID := resp.GetPayload().DeploymentID
+	// the new disk size
+	capacityInGB := int32(d.Get("capacity_in_gb").(int))
+
+	// get all the resources within the deployment and get the block device id
+	// from the deployment resource API
+	//NOTE: The block device id in the state file is different from the id in the
+	// deployment resources API. So iterating over all the resources to get the block device id
+	// that the deployment resource api will understand
+	depResp, err := apiClient.Deployments.GetDeploymentByIDUsingGET(
+		deployments.NewGetDeploymentByIDUsingGETParams().WithDepID(strfmt.UUID(deploymentID)).
+			WithExpandResources(withBool(true)))
+	if err != nil {
+		return err
+	}
+
+	deploymentResources := depResp.Payload.Resources
+	var resourceID strfmt.UUID
+	for _, resource := range deploymentResources {
+		propertiesMap := make(map[string]interface{})
+		propertiesMap = resource.Properties.(map[string]interface{})
+		// The block device id in the state file should match with the resource id
+		// in the deployment resource API response
+		if propertiesMap["resourceId"] == id {
+			resourceID = resource.ID
+		}
+	}
+	inputMap := make(map[string]interface{})
+	inputMap["capacityGb"] = capacityInGB
+	resourceActionRequest := models.ResourceActionRequest{
+		ActionID: "Cloud.vSphere.Disk.Disk.Resize",
```

Comment:
ActionID is hard coded to vSphere specific. It will fail when you run it against other cloud specific disks.
Instead the ActionID should be read from the list of available actions on the disk resource and pick the right one for resize.

I recommend to check if the disk resize action is available first, and throw an error if it cannot be resized. If the action is available, then proceed with resize operation.